### ISSUE TITLE
Conflicting report for ComparingThisWithNull in semantically similar code

### ISSUE THREAD

**Issue by csrma** [google/error-prone#5074]
The following code is _correctly_ flagged for `ComparingThisWithNull`.

```
class BuggyCode {

    public boolean checkNull() {
        return this == null;  // BUG: ComparingThisWithNull
    }

    public boolean checkNotNull() {
        return this != null;  // BUG: ComparingThisWithNull
    }

    public boolean[] showBug() {
        return new boolean[]{checkNull(), checkNotNull()};
    }
}
```
The error log for this is:
```
warning: [DefaultPackage] Java classes shouldn't use default package
class BuggyCode {
^
    (see https://errorprone.info/bugpattern/DefaultPackage)
demo.java:4: error: [ComparingThisWithNull] this == null is always false, this != null is always true
        return this == null;  // BUG: ComparingThisWithNull
                    ^
    (see https://errorprone.info/bugpattern/ComparingThisWithNull)
demo.java:8: error: [ComparingThisWithNull] this == null is always false, this != null is always true
        return this != null;  // BUG: ComparingThisWithNull
                   ^
    (see https://errorprone.info/bugpattern/ComparingThisWithNull)
2 errors
1 warning
```

However, when executing its semantically equivalent code as follows, `ComparingThisWithNull` is not reported. Although this is also comparing `this` with `null` through variable `a`.

```
class BuggyCode {

    public boolean checkNull() {
        Object a = this;
        return a == null;  // Mutated: Comparing 'this' with 'null' replaced with 'a' with 'null'
    }

    public boolean checkNotNull() {
        Object a = this;
        return a != null;  // Mutated: Comparing 'this' with 'null' replaced with 'a' with 'null'
    }

    public boolean[] showBug() {
        return new boolean[]{checkNull(), checkNotNull()};
    }
}
```
The error log for this is:
```
warning: [DefaultPackage] Java classes shouldn't use default package
class BuggyCode {
^
    (see https://errorprone.info/bugpattern/DefaultPackage)
1 warning
```
ErrorProne version: `2.36.0`

Is this rule implicitly assuming that `this` will never be stored in any variable? Does it only check for literal `this` keyword only?

**Comment by graememorgan** [comment]
> Does it only check for literal this keyword only?

Yes. Did you see a real world bug where `this` was compared to `null` via an intermediate variable?

**Comment by graememorgan** [comment]
This is roughly the same issue as #5076. We'd need to do data flow analysis, which is more in the realm of checker.
